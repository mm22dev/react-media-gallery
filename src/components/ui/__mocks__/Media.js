const Media = () => <div className='media-mock'></div>

Media.displayName = 'MediaMock'

export default Media
