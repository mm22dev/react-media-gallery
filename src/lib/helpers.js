const numbConv = cssProp => Number.isNaN(parseFloat(cssProp)) ? 0 : parseFloat(cssProp)

const isNodeEl = el => el !== null && typeof el === 'object' && 'nodeType' in el

export const S = (selector, ancestorEl = null) =>
  isNodeEl(ancestorEl)
    ? ancestorEl.querySelectorAll(selector).length > 0
      ? ancestorEl.querySelectorAll(selector)
      : null
    : document.querySelectorAll(selector).length > 0
      ? document.querySelectorAll(selector)
      : null

export const gcs = el => window.getComputedStyle(el)

export const cloneObj = obj => JSON.parse(JSON.stringify(obj))

export const getHeight = el => {
  if (el !== null && isNodeEl(el)) {
    const height = el.offsetHeight
    const borderTopWidth = gcs(el).getPropertyValue('border-top-width')
    const borderBottomWidth = gcs(el).getPropertyValue('border-bottom-width')
    const paddingTop = gcs(el).getPropertyValue('padding-top')
    const paddingBottom = gcs(el).getPropertyValue('padding-bottom')
    return height - numbConv(borderBottomWidth) - numbConv(borderTopWidth) - numbConv(paddingTop) - numbConv(paddingBottom)
  } else {
    return 0
  }
}

export const getWidth = el => {
  if (el !== null && isNodeEl(el)) {
    const width = el.offsetWidth
    const borderLeftWidth = gcs(el).getPropertyValue('border-left-width')
    const borderRightWidth = gcs(el).getPropertyValue('border-left-width')
    const paddingLeft = gcs(el).getPropertyValue('padding-left')
    const paddingRight = gcs(el).getPropertyValue('padding-right')
    return width - numbConv(borderRightWidth) - numbConv(borderLeftWidth) - numbConv(paddingLeft) - numbConv(paddingRight)
  } else {
    return 0
  }
}

export const identityFn = f => f

export const emptyQueueObject = {
  src: null,
  captionText: null,
  width: null,
  height: null,
  type: null,
  object: null,
  options: null
}

export const emptyMediaContent = {
  tagFamily: null,
  innerProps: {}
}

export const emptySpringSettings = {
  from: {},
  to: {},
  config: {},
  immediate: false,
  onRest: null
}

export const emptyActionReducer = {
  newState: null,
  callback: null
}

export const pixelify = size => `${size}px`

export const depixelify = size => +size.replace(/[^\d.-]/g, '')

export const hide = elementsArray => elementsArray.map(el => { el.style = { ...el.style, display: 'none' } })

export const show = (elementsArray, displayVariant = null) => elementsArray.map(el => { el.style = { ...el.style, display: (displayVariant !== null) ? displayVariant : 'block' } })

export const fadeOut = (elementsArray, duration = 400) => elementsArray.map(el => { el.fade = duration === 0 ? { inProp: false, immediate: true } : { inProp: false, duration: duration, immediate: false } })

export const fadeIn = (elementsArray, duration = 400) => elementsArray.map(el => { el.fade = duration === 0 ? { inProp: true, immediate: true } : { inProp: true, duration: duration, immediate: false } })

export class PRTError extends Error {
  constructor (message) {
    super(message)
    this.name = 'PopReactrox Warning'
  }
}
