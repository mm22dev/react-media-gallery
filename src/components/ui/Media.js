import React from 'react'
import PropTypes from 'prop-types'
import { connectWithPopupSettings } from '../../providers/PopupSettings'
import { Fade } from '../_common'
import initData from '../../data/popupDefaultData.json'
import { identityFn, gcs, depixelify } from '../../lib/helpers'
import C from '../../actions/constants'

const MediaUI = ({ wrapper, content, structureSettings, actionPerformer, popupSettings }) => {
  const fadeProps = {
    inProp: false,
    immediate: true,
    ...(wrapper.applyFadeIn && {
      inProp: true,
      duration: popupSettings.fadeSpeed,
      immediate: false,
      onFadeComplete: () => actionPerformer(C.ON_MEDIAWRAPPER_FADE_COMPLETE.req, structureSettings)
    })
  }

  const onLoad = ev => {
    actionPerformer(
      C.ON_MEDIA_CONTENT_STATUS_CHANGE.req,
      'loaded',
      {
        dimensions: {
          width: depixelify(gcs(ev.currentTarget).getPropertyValue('width')),
          height: depixelify(gcs(ev.currentTarget).getPropertyValue('height'))
        },
        prevStatus: content.status
      }
    )
  }

  const mediaProps = {
    ...content.elToRender.innerProps,
    style: {
      ...(content.elToRender.innerProps && { ...content.elToRender.innerProps.style }),
      ...content.style
    },
    onLoad: content.status === 'loading' ? ev => onLoad(ev) : identityFn
  }

  return (
    <Fade { ...fadeProps }>
      <div className="pic" style={wrapper.style}>{
        content.elToRender.tagFamily
          ? React.createElement(content.elToRender.tagFamily, mediaProps)
          : null
      }</div>
    </Fade>
  )
}

MediaUI.propTypes = {
  wrapper: PropTypes.object,
  content: PropTypes.object,
  actionPerformer: PropTypes.func,
  popupSettings: PropTypes.object
}

MediaUI.defaultProps = {
  wrapper: initData.renderSettings.popup.childs.media.wrapper,
  content: initData.renderSettings.popup.childs.media.content,
  actionPerformer: identityFn
}

const Media = connectWithPopupSettings(MediaUI)
Media.displayName = 'Media'
export default Media
