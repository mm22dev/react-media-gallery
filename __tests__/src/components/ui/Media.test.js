import Media from '../../../../src/components/ui/Media'
import toJSON from 'enzyme-to-json'
import { compose } from 'redux'
import defaultPopupSettings from '../../../../src/data/popupSettings.json'
import { identityFn } from '../../../../src/lib/helpers'

const { shallow, mount } = Enzyme

describe('<Media /> Component', () => {
  let { popupSettings } = defaultPopupSettings
  let wrapper

  describe('Rendering UI', () => {
    it('Renders correctly', () => {
      compose(expect, toJSON, shallow)(<Media />).toMatchSnapshot()
    })

    describe('Rendering wrapper', () => {
      it('Callback to unlock Popup is called on fadeIn animation end', done => {
        const mockedWrapper = {
          style: {},
          applyFadeIn: true
        }
        const _actionPerformer = jest.fn(action => {
          expect(action).toBe('onMediaWrapperFadeComplete')
          done()
        })
        wrapper = mount(<Media wrapper={mockedWrapper} actionPerformer={_actionPerformer} popupSettings={popupSettings} />)
      })
    })

    describe('Rendering content', () => {
      let mockedContent = {
        style: {},
        status: 'idle',
        elToRender: {
          tagFamily: null,
          innerProps: {}
        }
      }

      beforeAll(() => {
        wrapper = mount(<Media />)
      })

      it('No HTML elements are rendered if specifications are not provided', () => {
        expect(wrapper.find('.pic').isEmptyRender()).toBe(true)
      })

      it('HTML elements built on specifications are correctly rendered', () => {
        mockedContent.elToRender.tagFamily = 'div'
        wrapper.setProps({ content: mockedContent })
        expect(wrapper.find('.pic').isEmptyRender()).not.toBe(true)
      })

      describe('Providing a callback on media content loaded', () => {
        let _actionPerformer

        beforeAll(() => {
          mockedContent.elToRender.innerProps.key = '5c608e9d-c265-465b-8925-9c4f6d80a782'
          mockedContent.elToRender.innerProps.src = 'http://via.placeholder.com/350x150'
          mockedContent.elToRender.tagFamily = 'img'
        })

        it('Do not invokes callback if content status is idle', () => {
          mockedContent.status = 'idle'
          wrapper.setProps({ content: mockedContent })
          expect(wrapper.find('MediaUI .pic').childAt(0).props().onLoad).toEqual(identityFn)
        })

        it('Invokes callback if content status is loading', done => {
          _actionPerformer = jest.fn((action, mediaStatus, forwardingArgs) => {
            expect(action).toBe('onMediaContentStatusChange')
            expect(mediaStatus).toBe('loaded')
            expect(forwardingArgs.dimensions).toHaveProperty('width')
            expect(forwardingArgs.dimensions).toHaveProperty('height')
            expect(forwardingArgs.prevStatus === 'loading' || forwardingArgs.prevStatus === 'resize').toBeTruthy()
            done()
          })
          mockedContent.status = 'loading'
          wrapper.setProps({ content: mockedContent, actionPerformer: _actionPerformer })
          wrapper.find('MediaUI .pic').childAt(0).simulate('load')

          mockedContent.status = 'resize'
          wrapper.setProps({ content: mockedContent, actionPerformer: _actionPerformer })
          wrapper.find('MediaUI .pic').childAt(0).simulate('load')
        })
      })
    })
  })
})
