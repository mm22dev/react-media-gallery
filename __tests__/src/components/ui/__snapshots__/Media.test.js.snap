// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Media /> Component Rendering UI Renders correctly 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
